### PR TITLE
[Tasks] Adds image captioning 

### PR DESCRIPTION
**Pull request by sayakpaul**
This PR adds a guide on image captioning introducing a brand new section for tasks: Multimodal :) 

The idea was originated by @NielsRogge and the PR builds on top of his great work [here](https://github.com/NielsRogge/Transformers-Tutorials/blob/master/GIT/Fine_tune_GIT_on_an_image_captioning_dataset.ipynb). 

An interactive Colab Notebook that was used to develop the task guide is available [here](https://colab.research.google.com/drive/1sLiSY2ixv52yqlw9EUW6BPwCq9XPLc9R?usp=sharing).

Not bad quality, eh?

image

Generated caption: 

> a drawing of a pink and blue pokemon

### CODE REVIEW

File: docs/source/en/tasks/image_captioning.mdx [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+
+```python
+from huggingface_hub import notebook_login
+
+notebook_login()
+```
+
+## Load the Pokemons dataset
+
+Use the 🤗 Dataset library to load a dataset that consists of {image-caption} pairs. If you are looking for a more customizable treatment of the dataset, you can follow [this notebook](https://github.com/NielsRogge/Transformers-Tutorials/blob/master/GIT/Fine_tune_GIT_on_an_image_captioning_dataset.ipynb).
```
Comment (sgugger) [inline docs/source/en/tasks/image_captioning.mdx:43]: Why the curly brackets?

File: docs/source/en/tasks/image_captioning.mdx [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+Split the dataset’s train split into a train and test set with the [`train_test_split()`](https://huggingface.co/docs/datasets/main/en/package_reference/main_classes#datasets.Dataset.train_test_split) method:
+
+
+```python
+ds = ds["train"].train_test_split(test_size=0.1)
+train_ds = ds["train"]
+test_ds = ds["test"]
+```
+
+Visualize a couple of samples from the training set.
```
Comment (sgugger) [inline docs/source/en/tasks/image_captioning.mdx:78]:
```suggestion
Let's visualize a couple of samples from the training set. 
```

File: docs/source/en/tasks/image_captioning.mdx [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+from transformers import AutoModelForCausalLM
+
+model = AutoModelForCausalLM.from_pretrained(checkpoint)
+```
+
+## Evaluate
+
+For evaluating image captioning models, the typical metrics are [Rouge Score](https://huggingface.co/spaces/evaluate-metric/rouge) and [Word Error Rate](https://huggingface.co/spaces/evaluate-metric/wer). For this guide, you will use the Word Error Rate (WER). 
+
+Use the 🤗 Evaluate library to do so. For potential limitations and other gotchas of the WER, refer to [this guide](https://huggingface.co/spaces/evaluate-metric/wer).
```
Comment (sgugger) [inline docs/source/en/tasks/image_captioning.mdx:152]:
```suggestion
We use the 🤗 Evaluate library to do so. For potential limitations and other gotchas of the WER, refer to [this guide](https://huggingface.co/spaces/evaluate-metric/wer). 
```

File: docs/source/en/tasks/image_captioning.mdx [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+    decoded_predictions = processor.batch_decode(predicted, skip_special_tokens=True)
+    wer_score = wer.compute(predictions=decoded_predictions, references=decoded_labels)
+    return {"wer_score": wer_score}
+```
+
+## Train!
+
+Now, you are ready to start fine-tuning the model. You will [🤗 Trainer](https://huggingface.co/docs/transformers/main/en/main_classes/trainer#transformers.Trainer) for this. 
+
+First, define the training arguments using [`TrainingArguments`](https://huggingface.co/docs/transformers/v4.26.0/en/main_classes/trainer#transformers.TrainingArguments). Then pass them along with the datasets and the model to 🤗 Trainer. To start training, simply call `train()` on the `Trainer` object.
```
Comment (sgugger) [inline docs/source/en/tasks/image_captioning.mdx:175]:
```suggestion
First, define the training arguments using [`TrainingArguments`]. Then pass them along with the datasets and the model to 🤗 Trainer. To start training, simply call [`~Trainer.train`] on the [`Trainer`] object.
```

File: docs/source/en/tasks/image_captioning.mdx [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+
+trainer = Trainer(
+    model=model,
+    args=training_args,
+    train_dataset=train_ds,
+    eval_dataset=test_ds,
+    compute_metrics=compute_metrics,
+)
+
+trainer.train()
```
Comment (sgugger) [inline docs/source/en/tasks/image_captioning.mdx:211]: You should split this in several code blocks so it's split in several cells in the notebook. You can take parts of the texts on top to put between the code blocks!

File: docs/source/en/tasks/image_captioning.mdx [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+```
+
+You should expect to see the training loss going down smoothly as training progresses.
+
+Once training is completed, share your model to the Hub with the [`push_to_hub()`](https://huggingface.co/docs/transformers/main/en/main_classes/trainer#transformers.Trainer.push_to_hub) method so everyone can use your model:
```
Comment (sgugger) [inline docs/source/en/tasks/image_captioning.mdx:216]:
```suggestion
Once training is completed, share your model to the Hub with the [`~Trainer.push_to_hub`] method so everyone can use your model:
```

**Review by sgugger — APPROVED** [review]
Thanks for adding this! I have a couple of comments regarding links, but otherwise LGTM!

File: docs/source/en/tasks/image_captioning.mdx (hunk 1 above)
Comment (stevhliu) [inline docs/source/en/tasks/image_captioning.mdx:43]: > If you are looking for a more customizable treatment of the dataset

I'd also be more specific about what you're describing here, so users know exactly what to expect when they navigate away. You can say something like: To create your own image captioning dataset in PyTorch, you can follow this notebook.

File: docs/source/en/tasks/image_captioning.mdx [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+We encourage you to log in to your Hugging Face account so you can upload and share your model with the community. When prompted, enter your token to log in:
+
+
+```python
+from huggingface_hub import notebook_login
+
+notebook_login()
+```
+
+## Load the Pokemons dataset
```
Comment (stevhliu) [inline docs/source/en/tasks/image_captioning.mdx:41]:
```suggestion
## Load the Pokémon BLIP captions dataset
```

File: docs/source/en/tasks/image_captioning.mdx [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+Use the 🤗 Dataset library to load a dataset that consists of {image-caption} pairs. If you are looking for a more customizable treatment of the dataset, you can follow [this notebook](https://github.com/NielsRogge/Transformers-Tutorials/blob/master/GIT/Fine_tune_GIT_on_an_image_captioning_dataset.ipynb). 
+
+
+```python
+from datasets import load_dataset
+
+ds = load_dataset("lambdalabs/pokemon-blip-captions")
+ds
+```
+```bash
```
Comment (stevhliu) [inline docs/source/en/tasks/image_captioning.mdx:52]: I think this code block showing `ds` can be merged with the block above.

File: docs/source/en/tasks/image_captioning.mdx [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+DatasetDict({
+    train: Dataset({
+        features: ['image', 'text'],
+        num_rows: 833
+    })
+})
+```
+
+Notice that the dataset has got two features namely `image` and `text` which are self-explanatory.
```
Comment (stevhliu) [inline docs/source/en/tasks/image_captioning.mdx:61]:
```suggestion
The dataset has two features, `image` and `text`.
```

File: docs/source/en/tasks/image_captioning.mdx [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+        caption = captions[i]
+        caption = "\n".join(wrap(caption, 12))
+        plt.title(caption)
+        plt.imshow(images[i])
+        plt.axis("off")
+
+
+sample_images_to_visualize = [np.array(train_ds[i]["image"]) for i in range(5)]
+sample_captions = [train_ds[i]["text"] for i in range(5)]
+plot_images(sample_images_to_visualize, sample_captions)
```
Comment (stevhliu) [inline docs/source/en/tasks/image_captioning.mdx:100]: So cool! 🤩

File: docs/source/en/tasks/image_captioning.mdx [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+```
+    
+<div class="flex justify-center">
+    <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/transformers/tasks/sample_training_images_image_cap.png" alt="Sample training images"/>
+</div>
+
+## Preprocess the dataset
+
+Since the dataset has two different modalities (image and text), the pre-processing pipeline will include pre-processing of the images as well the text (the captions).
```
Comment (stevhliu) [inline docs/source/en/tasks/image_captioning.mdx:109]:
```suggestion
Since the dataset has two modalities (image and text), the pre-processing pipeline will preprocess images and the captions.
```

File: docs/source/en/tasks/image_captioning.mdx [hunk 12]
```diff
@@ -0,0 +1,261 @@
...
+    inputs = processor(images=images, text=captions, padding="max_length")
+    inputs.update({"labels": inputs["input_ids"]})
+    return inputs
+
+
+train_ds.set_transform(transforms)
+test_ds.set_transform(transforms)
+```
+
+With the datasets ready, proceed to setting up the model for fine-tuning.
```
Comment (stevhliu) [inline docs/source/en/tasks/image_captioning.mdx:135]:
```suggestion
With the dataset ready, you can now set up the model for fine-tuning. 
```

File: docs/source/en/tasks/image_captioning.mdx [hunk 13]
```diff
@@ -0,0 +1,261 @@
...
+
+```python
+from transformers import AutoModelForCausalLM
+
+model = AutoModelForCausalLM.from_pretrained(checkpoint)
+```
+
+## Evaluate
+
+For evaluating image captioning models, the typical metrics are [Rouge Score](https://huggingface.co/spaces/evaluate-metric/rouge) and [Word Error Rate](https://huggingface.co/spaces/evaluate-metric/wer). For this guide, you will use the Word Error Rate (WER).
```
Comment (stevhliu) [inline docs/source/en/tasks/image_captioning.mdx:150]:
```suggestion
Image captioning models are typically evaluated with the [Rouge Score](https://huggingface.co/spaces/evaluate-metric/rouge) or [Word Error Rate](https://huggingface.co/spaces/evaluate-metric/wer). For this guide, you will use the Word Error Rate (WER). 
```

File: docs/source/en/tasks/image_captioning.mdx [hunk 14]
```diff
@@ -0,0 +1,261 @@
...
+    predicted = logits.argmax(-1)
+    decoded_labels = processor.batch_decode(labels, skip_special_tokens=True)
+    decoded_predictions = processor.batch_decode(predicted, skip_special_tokens=True)
+    wer_score = wer.compute(predictions=decoded_predictions, references=decoded_labels)
+    return {"wer_score": wer_score}
+```
+
+## Train!
+
+Now, you are ready to start fine-tuning the model. You will [🤗 Trainer](https://huggingface.co/docs/transformers/main/en/main_classes/trainer#transformers.Trainer) for this.
```
[the remark this answers is not in the document]
Comment (stevhliu) [reply docs/source/en/tasks/image_captioning.mdx:173]: You will **use** the [`Trainer`] for this. :)

File: docs/source/en/tasks/image_captioning.mdx [hunk 15]
```diff
@@ -0,0 +1,261 @@
...
+    args=training_args,
+    train_dataset=train_ds,
+    eval_dataset=test_ds,
+    compute_metrics=compute_metrics,
+)
+
+trainer.train()
+```
+
+You should expect to see the training loss going down smoothly as training progresses.
```
Comment (stevhliu) [inline docs/source/en/tasks/image_captioning.mdx:214]:
```suggestion
You should see the training loss drop smoothly as training progresses.
```

File: docs/source/en/tasks/image_captioning.mdx [hunk 16]
```diff
@@ -0,0 +1,261 @@
...
+Prepare image for the model.
+
+```python
+device = "cuda" if torch.cuda.is_available() else "cpu"
+
+inputs = processor(images=image, return_tensors="pt").to(device)
+pixel_values = inputs.pixel_values
+```
+
+Perform inference and decode the predictions.
```
Comment (stevhliu) [inline docs/source/en/tasks/image_captioning.mdx:250]:
```suggestion
Call [`generate`] and decode the predictions. 
```

File: docs/source/en/tasks/image_captioning.mdx [hunk 17]
```diff
@@ -0,0 +1,261 @@
...
+
+Perform inference and decode the predictions. 
+
+```python
+generated_ids = model.generate(pixel_values=pixel_values, max_length=50)
+generated_caption = processor.batch_decode(generated_ids, skip_special_tokens=True)[0]
+print(generated_caption)
+```
+```bash
+a drawing of a pink and blue pokemon
```
Comment (stevhliu) [inline docs/source/en/tasks/image_captioning.mdx:258]: I think the output will also look better combined with the code block above.

File: docs/source/en/tasks/image_captioning.mdx [hunk 18]
```diff
@@ -0,0 +1,261 @@
...
+```
+
+Looks like the fine-tuned model generated a good caption.
```
Comment (stevhliu) [inline docs/source/en/tasks/image_captioning.mdx:261]:
```suggestion
Looks like the fine-tuned model generated a pretty good caption!
```

File: docs/source/en/tasks/image_captioning.mdx [hunk 19]
```diff
@@ -0,0 +1,261 @@
...
+You should expect to see the training loss going down smoothly as training progresses.
+
+Once training is completed, share your model to the Hub with the [`push_to_hub()`](https://huggingface.co/docs/transformers/main/en/main_classes/trainer#transformers.Trainer.push_to_hub) method so everyone can use your model:
+
+
+```python
+trainer.push_to_hub()
+```
+
+## Inference
```
Comment (stevhliu) [inline docs/source/en/tasks/image_captioning.mdx:223]: I think it'd also be nice to show you can use the `ImageToText` pipeline

**Review by stevhliu — COMMENTED** [review]
Super cool, thanks so much for adding this! The main feedback is to consider showing you can also use it in a `pipeline`. 🙂

File: docs/source/en/tasks/image_captioning.mdx (hunk 1 above)
Comment (MKhalusova) [reply docs/source/en/tasks/image_captioning.mdx:43]: +1 to this. I was not sure what the "more customizable treatment of the dataset" meant until I went to check the notebook.

File: docs/source/en/tasks/image_captioning.mdx [hunk 20]
```diff
@@ -0,0 +1,261 @@
...
+        num_rows: 833
+    })
+})
+```
+
+Notice that the dataset has got two features namely `image` and `text` which are self-explanatory. 
+
+<Tip>
+
+Many image captioning datasets contain multiple captions per image. In those cases, a common strategy is to randomly sample a caption amongst the available ones during training.
```
Comment (MKhalusova) [inline docs/source/en/tasks/image_captioning.mdx:65]: To make sure I understand correctly, does it mean that I have to sample? Or can I pass multiple captions to the model and let it sample by specifying some parameters?

**Review by MKhalusova — COMMENTED** [review]
Thank you for the guide! Really nice example too :) My main nitpick is something that @stevhliu pointed out already - it would be nice to show inference with the corresponding pipeline.

File: docs/source/en/tasks/image_captioning.mdx (hunk 20 above)
Comment (sayakpaul) [reply docs/source/en/tasks/image_captioning.mdx:65]: This process should be handled by the users in the data pipeline.

File: docs/source/en/tasks/image_captioning.mdx (hunk 19 above)
Comment (sayakpaul) [reply docs/source/en/tasks/image_captioning.mdx:223]: See: https://github.com/huggingface/transformers/issues/21514

File: docs/source/en/tasks/image_captioning.mdx (hunk 1 above)
Comment (sayakpaul) [reply docs/source/en/tasks/image_captioning.mdx:43]: Motivated from https://keras.io/guides/keras_cv/generate_images_with_stable_diffusion/

File: docs/source/en/tasks/image_captioning.mdx (hunk 3 above)
Comment (sayakpaul) [reply docs/source/en/tasks/image_captioning.mdx:152]: @sgugger, 

Maintaining single first-person nouns here following @MKhalusova's advice from other PRs.

File: docs/source/en/tasks/image_captioning.mdx (hunk 8 above)
Comment (sayakpaul) [reply docs/source/en/tasks/image_captioning.mdx:52]: I didn't want to do it because the following cell is not of type "`py`". 

Users might directly copy-paste code from these cells. If we merged two cells together where one cell is of type Python and another one is of type Bash, it would results in an error.

File: docs/source/en/tasks/image_captioning.mdx (hunk 17 above)
Comment (sayakpaul) [reply docs/source/en/tasks/image_captioning.mdx:258]: Same as https://github.com/huggingface/transformers/pull/21512/files#r1100943522

File: docs/source/en/tasks/image_captioning.mdx (hunk 3 above)
Comment (sgugger) [reply docs/source/en/tasks/image_captioning.mdx:155]: That's not the style of the doc and I read the imperative form as slightly aggressive, but your choice.

Comment (sayakpaul) [reply docs/source/en/tasks/image_captioning.mdx:155]: I'm retrospect, I think adding what you suggested (We use ...) won't hurt the tone. 

@MKhalusova WDYT?

Comment (MKhalusova) [reply docs/source/en/tasks/image_captioning.mdx:155]: It's ok to make exceptions to the rule. I would say stick to first-person for the most part, but if it feels forced, it's ok to use "we".

**Review by MKhalusova — APPROVED** [review]
Great work! LGTM :)
